src/USER-HSMA/hsma2d.cpp

```diff
@@ -0,0 +1,2056 @@
+		for (int j = -ly - 1; j <= ly + 1; j++)
+			for (int k = -lz; k <= lz; k++)
+			{
+				for (int m = 0; m < NSource; m++)
+				{
+					CX = Source[m][0] + i * Lx;
+					CY = Source[m][1] + j * Ly;
+					CZ = pow(-1, k) * Source[m][2] + k * Lz;
+					if (CX * CX + CY * CY + CZ * CZ <= Rs * Rs)
+					{				   
+						ImageCharge[number][0] = CX;
+						ImageCharge[number][1] = CY;
+						ImageCharge[number][2] = CZ;
+						ImageCharge[number][3] = Q[m];
+						ImageCharge[number][4] = k;
+						number++;
+					}
+				}
+			}
+	}
+
+	*ImageNumber = number;
+	return 1.0;
+}
+
+void HSMA2D::CalculateNearFieldAndZD(double** Top, double** TopZD, double** Down, double** DownZD, double ImageCharge[][5], int ImageNumber, double*** IntegralTop, double*** IntegralDown, int Nw, double Gamma, int IF_FMM_RightTerm, int S, double* AR, double Lx, double Ly, double Lz, double tolerance)
+{
+	if (IF_FMM_RightTerm)//Using FMM to calculate pairwise sum
+	{
+		double eps = tolerance;
+
+		/*            Set FMM parameters           */
+		int ns = ImageNumber;
+		int nt = (S * Nw) * (S * Nw) * 2;
+		double* source = (double*)malloc(3 * ns * sizeof(double));
+		double* target = (double*)malloc(3 * nt * sizeof(double));
+		double* charge = (double*)malloc(ns * sizeof(double));
+
+		for (int i = 0; i < ns; i++)
+		{
+			source[3 * i] = ImageCharge[i][0];
+			source[3 * i + 1] = ImageCharge[i][1];
+			source[3 * i + 2] = ImageCharge[i][2];
+			charge[i] = ImageCharge[i][3] * pow(Gamma, fabs(ImageCharge[i][4]) + 0.00);
+		}
+
+		for (int i = 0; i < S * Nw; i++)
+			for (int j = 0; j < S * Nw; j++)
+			{
+				target[3 * (i * (S * Nw) + j)] = IntegralTop[i][j][0];
+				target[3 * (i * (S * Nw) + j) + 1] = IntegralTop[i][j][1];
+				target[3 * (i * (S * Nw) + j) + 2] = IntegralTop[i][j][2];
+
+				target[3 * ((S * Nw) * (S * Nw) + i * (S * Nw) + j)] = IntegralDown[i][j][0];
+				target[3 * ((S * Nw) * (S * Nw) + i * (S * Nw) + j) + 1] = IntegralDown[i][j][1];
+				target[3 * ((S * Nw) * (S * Nw) + i * (S * Nw) + j) + 2] = IntegralDown[i][j][2];
+			}
+
+		double* pottarg = (double*)malloc(nt * sizeof(double));
+		double* gradtarg = (double*)malloc(3 * nt * sizeof(double));
+
+		int ier = 0;
+		lfmm3d_t_c_g_(&eps, &ns, source, charge, &nt, target, pottarg, gradtarg, &ier);
+
+		for (int i = 0; i < S * Nw; i++)
+			for (int j = 0; j < S * Nw; j++)
+			{
+				Top[i][j] = pottarg[i * (S * Nw) + j];
+				TopZD[i][j] = gradtarg[3 * (i * (S * Nw) + j) + 2];
+				Down[i][j] = pottarg[(S * Nw) * (S * Nw) + i * (S * Nw) + j];
+				DownZD[i][j] = gradtarg[3 * ((S * Nw) * (S * Nw) + i * (S * Nw) + j) + 2];
+			}
+		
+		free(source); free(target); free(charge); free(pottarg); free(gradtarg);
+	}
+	else
+	{
+		double Paramet1[ImageNumber];
+		for (int i = 0; i < ImageNumber; i++)
+		{
+			Paramet1[i] = ImageCharge[i][3] * pow(Gamma, fabs(ImageCharge[i][4]) + 0.00);
+		}
+		
+		#pragma omp parallel
+		{
+			int id = omp_get_thread_num();
+			int size = omp_get_num_threads();
```

Comment:
Please do not retrieve the number of threads directly, but synchronize this with the `nthreads` member in the `Comm` class, so that setting the number of threads via the package command or the kokkos flag will result in a consistent number of threads in use throughout LAMMPS.

Reply:
Change `omp_get_num_threads()` to `comm->nthreads`